aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py

```diff
@@ -0,0 +1,823 @@
+
+    Args:
+        credentials: list of VCRecords to check
+        schemas: list of schemas from the input_descriptors
+    Return:
+        Sequence of filtered VCRecord
+
+    """
+    result = []
+    for tmp_cred in credentials:
+        applicable = False
+        for tmp_schema in schemas:
+            applicable = await credential_match_schama(
+                credential=tmp_cred, schema_id=tmp_schema.uri
+            )
+            if tmp_schema.required and not applicable:
+                break
+        if applicable:
+            result.append(tmp_cred)
+    return result
+
+
+async def credential_match_schama(credential: VCRecord, schema_id: str) -> bool:
+    """
+    Credential matching by schema.
+
+    Used by filter_schema to check if credential.schema_ids or credential.types
+    matched with schema_id
+
+    Args:
+        credential: VCRecord to check
+        schema_id: schema uri to check
+    Return:
+        bool
+    """
+    for cred_schema_id in credential.schema_ids:
+        if cred_schema_id == schema_id:
+            return True
+    for cred_type in credential.types:
+        if cred_type == schema_id:
+            return True
+    return False
+
+
+async def apply_requirements(req: Requirement, credentials: Sequence[VCRecord]) -> dict:
+    """
+    Apply Requirement.
+
+    Args:
+        req: Requirement
+        credentials: Sequence of credentials to check against
+    Return:
+        dict of input_descriptor ID key to list of credential_json
+    """
+    result = {}
+    descriptor_list = []
+    if not req.input_descriptors:
+        descriptor_list = []
+    else:
+        descriptor_list = req.input_descriptors
+    for tmp_descriptor in descriptor_list:
+        filtered_by_schema = await filter_schema(
+            credentials=credentials, schemas=tmp_descriptor.schemas
+        )
+        filtered = await filter_constraints(
+            constraints=tmp_descriptor.constraint, credentials=filtered_by_schema
+        )
+        if len(filtered) != 0:
+            result[tmp_descriptor._id] = filtered
+
+    if len(descriptor_list) != 0:
+        if await is_len_applicable(req, len(result)):
+            return result
+        return {}
+
+    nested_result = []
+    tmp_dict = {}
+    # recursion logic for nested requirements
+    for tmp_req in req.nested_req:
+        tmp_result = await apply_requirements(tmp_req, credentials)
+        if tmp_result == {}:
+            continue
+
+        for tmp_desc_id in tmp_result.keys():
+            tmp_creds_list = tmp_result.get(tmp_desc_id)
+            for tmp_cred in tmp_creds_list:
+                if await trim_tmp_id(tmp_cred.given_id) not in tmp_dict:
+                    tmp_dict[await trim_tmp_id(tmp_cred.given_id)] = {}
+                tmp_dict[await trim_tmp_id(tmp_cred.given_id)][
+                    tmp_desc_id
+                ] = tmp_cred.given_id
+
+        if len(tmp_result.keys()) != 0:
+            nested_result.append(tmp_result)
+
```

Comment:
I'm sorry but I have no idea what's happening here. Could you maybe add some comments / more descriptive names or restructure a bit?